stdlib/asm/math/poly512.masm

```diff
@@ -0,0 +1,6682 @@
+        exec.mod_div
+    end
+
+    push.12289
+    swap
+    sub
+end
+
+# Given four elements on stack, this routine negates those, using modular arithmetic
+# over Falcon Digital Signature Algorithm's prime field = 12289.
+#
+# All this routine does is
+#
+# b`i` = (0 - a`i`) % Q
+#   = Q - a`i` % Q | Q = 12289 & i ∈ [0..4)
+#
+# Input stack state
+#
+# [a0, a1, a2, a3, ...]
+#
+# Output stack state looks like
+#
+# [b0, b1, b2, b3 ...] | b`i` ∈ [0..12289)
+proc.neg_word
+    exec.neg
+
+    swap
+    exec.neg
+    swap
+
+    movup.2
+    exec.neg
+    movdn.2
+
+    movup.3
+    exec.neg
+    movdn.3
+end
+
+# Given a field element, this routine does centered reduction using Miden VM
+# prime ( say Q ) and then reduces it using Falcon Post Quantum Digital 
+# Signature Algorithm prime ( say Q' )
+#
+# Q = 2 ^ 64 - 2 ^ 32 + 1
+# Q' = 12289
+#
+# Expected stack state
+#
+# [a, ...]
+#
+# All this routine does is
+#
+# if a > (Q >> 1):
+#   b = (a - Q) % Q'
+# else:
+#   b = a % Q'
+#
+# Final stack state looks like
+#
+# [b, ...]
+proc.reduce
+    dup
+    push.9223372034707292160
+    gt
+
+    if.true
+        push.12289
+        exec.mod_div
+
+        dup
+        push.7002
+        gte
+
+        if.true
+            sub.7002
+        else
+            push.7002
+            swap
+            sub
+
+            push.12289
+            swap
+            sub
+        end
+    else
+        push.12289
+        exec.mod_div
+    end
+end
+
+# Reduces four consecutive elements living on stack top using `reduce` routine ( defined above )
+#
+# Expected stack state
+#
+# [a0, a1, a2, a3, ...]
+#
+# What this routine does is
+#
+# b`i` = reduce(a`i`)
+#
+# Final stack state looks like
+#
+# [b0, b1, b2, b3, ...]
+proc.reduce_word
+    exec.reduce
+
+    swap
+    exec.reduce
+    swap
+
+    movup.2
+    exec.reduce
+    movdn.2
+
+    movup.3
+    exec.reduce
+    movdn.3
+end
+
+# Given two polynomials of degree 512 on stack as absolute memory addresses,
+# this routine computes polynomial multiplication, using NTT and iNTT.
+#
+# Imagine, two polynomials are f, g
+#
+# h = f . g, can be computed using
+#
+# iNTT(NTT(f) * NTT(g))
+#
+# Note, * -> element wise multiplication of polynomial coefficients in NTT domain
+#
+# Input stack state :
+#
+# [f_addr0, f_addr1, ..., f_addr127, g_addr0, g_addr1, ..., g_addr127]
+#
+# {f, g}_addr`i` -> {f, g}[ (i << 2) .. ((i+1) << 2) ), address holding four consecutive coefficients
+#
+# Output stack state :
+#
+# [h_addr0, h_addr1, ..., h_addr127]
+#
+# h_addr`i` is the address of four elements | h[ (i << 2) .. ((i+1) << 2) ) & i ∈ [0..128)
+#
+# Note, input memory addresses are considered to be read-only, they are not mutated.
+export.mul_zq.128
+    exec.ntt512::forward
+
+    push.0.0.0.0
+
+    movup.4
+    loadw.mem
+    storew.local.0
+
+    movup.4
+    loadw.mem
+    storew.local.1
+
+    movup.4
+    loadw.mem
+    storew.local.2
+
+    movup.4
+    loadw.mem
+    storew.local.3
```

Comment:
I think what these and following instructions do is basically a memcopy from one address into another. If so, it might be worth creating a separate procedure to handle this (would improve code readability).

Reply:
Yes, that's exactly what I do in commit a97f21c. It reduces lines of code to ~0.6k from ~6.7k.